addons/io_hubs_addon/components/operators.py

```diff
@@ -670,6 +671,207 @@ def invoke(self, context, event):
+
+            # The layer for the lightmap needs to be the active one before lightmap packing
+            obj_uv_layers.active = obj_uv_layers[LIGHTMAP_LAYER_NAME]
+            # Set the object as selected in object mode
+            obj.select_set(True)
+
+        # run UV lightmap packing on all selected objects
+        bpy.ops.object.mode_set(mode='EDIT')
+        bpy.ops.mesh.select_all(action='SELECT')
+        # bpy.ops.uv.lightmap_pack()
+        bpy.ops.uv.smart_project(island_margin=LIGHTMAP_UV_ISLAND_MARGIN)
+        bpy.ops.object.mode_set(mode='OBJECT')
+        # Deselct the objects again to return without changing the scene
+        for obj in mesh_objs:
+            obj.select_set(False)
+        # Update the view layer so all parts take notice of the changed UV layout
+        bpy.context.view_layer.update()
+
+        return{'FINISHED'}
+
+    def execute(self, context):
+        # Check selected objects
+        selected_objects = bpy.context.selected_objects
+
+        # filter mesh objects and others
+        mesh_objs, other_objs = [], []
+        for ob in selected_objects:
+            (mesh_objs if ob.type == 'MESH' else other_objs).append(ob)
+            # Remove all objects from selection so we can easily re-select subgroups later
+            ob.select_set(False)
+
+        # for ob in other_objs:
+            # Remove non-mesh objects from selection to ensure baking will work
+            # ob.select_set(False)
+
+        # Gather all materials on the selected objects
+        # materials = []
+        # Dictionary that stores which object has which materials so we can group them later
+        material_object_associations = {}
+        for obj in mesh_objs:
+            if len(obj.material_slots) >= 1:
+                # TODO: Make more efficient
+                for slot in obj.material_slots:
+                    if slot.material is not None:
+                        mat = slot.material
+                        if mat not in material_object_associations:
+                            # materials.append(mat)
+                            material_object_associations[mat] = []
+                        material_object_associations[mat].append(obj)
+            else:
+                # an object without materials should not be selected when running the bake operator
+                print("Object " + obj.name + " does not have material slots, removing from set")
+                obj.select_set(False)
+                mesh_objs.remove(obj)
+
+        print(material_object_associations.items())
+        # Set up the UV layer structure and auto-unwrap optimized for lightmaps
+        visited_objects = set()
+        for mat, obj_list in material_object_associations.items():
+            for ob in visited_objects:
+                if ob in obj_list:
+                    obj_list.remove(ob)
+            self.create_uv_layouts(context, obj_list)
+            for ob in obj_list:
+                visited_objects.add(ob)
```

Comment:
This visited_objects logic seems wrong. I spend 1h thinking about it, I hope my example will be clear.

We agree that an object can have several materials, each material with their own image texture, the lightmap uvmap is shared for those materials on an object.

Example:
obj1 with alu and glass materials
obj2 with alu
obj3 with glass
obj4 with rock

We handle alu material first, create uvmap for obj1 obj2 on same image texture for alu.
Then we handle glass material, we don't recreate the uvmap for obj1 because of your visited_objects, we create the uvmap for obj3. The faces for obj1 and obj3 may end up overlapping on the same image texture for glass.

What we need to do is create the uvmap for obj1 obj2 obj3 with the same smart uv project.
For example with alu, glass and rock materials, 
iterate over materials, iterating over alu objects, if an object has other materials, include all objects of those materials to do the uv smart project. Mark alu and glass as done. Continue iterating over materials, skip glass because it's done, then handle rock material.

Reply:
You are totally right, it needs to work by material and not by object, should be adressed by https://github.com/Hubs-Foundation/hubs-blender-exporter/pull/281/commits/a91a5b43f363b1558e9f551c34f62f05eb42f991